lib/ui/worker/upgrade.dart

```diff
@@ -266,3 +305,181 @@ extension Rfc822ToDateTime on DateTime {
+            if (ourSecond == theirSecond) {
+              // If the second parts are equal, then the deeper parts are
+              // different, which isn't considered a critical release.
+              //
+              // Example: `0.1.0-alpha.1` -> `0.1.0-alpha.1.1` => `false`.
+              // Example: `0.1.0-alpha.1.1.2` -> `0.1.0-alpha.1.1.3` => `false`.
+              // Example: `0.1.0-alpha.1.5` -> `0.1.0-alpha.1.6` => `false`.
+              return false;
+            } else {
+              // If our second pre-release part is lower than the second
+              // pre-release part of the other version, then this is critical.
+              //
+              // Example: `0.1.0-alpha.1` -> `0.1.0-alpha.2` => `true`.
+              return ourSecond.compareTo(theirSecond) < 0;
+            }
+          }
+        } else {
+          // If our first pre-release part is lower than the first pre-release
+          // part of the other version, then this is critical.
+          //
+          // Example: `0.1.0-alpha` -> `0.1.0-beta` => `true`.
+          return ourFirst.compareTo(theirFirst) < 0;
+        }
+      }
+    }
+
+    // If the compared versions contain no pre-releases, then proceed comparing
+    // major/minor numbers.
+    //
+    // Differences in major number are always considered critical.
+    //
+    // Example: `1.0.0` -> `2.0.0` => `true`.
+    if (major < other.major) {
+      return true;
+    }
+
+    if (major == 0) {
+      // If major is `0`, then differences in minor number are considered
+      // critical.
+      //
+      // Example: `0.1.0` -> `0.2.0` => `true`.
+      if (minor < other.minor) {
+        return true;
+      }
+    }
+
+    // Otherwise this isn't a critical release.
+    //
+    // Example: `0.1.0` -> `0.1.1` => `false`.
+    // Example: `1.0.0` -> `1.0.1` => `false`.
+    // Example: `1.0.0` -> `1.2.3` => `false`.
+    return false;
+  }
+}
+
+/// Extension adding [Version]s parsing with hyphens in pre-releases parsed as
+/// separate parts.
+extension VersionExtension on Version {
+  /// Returns the [Version] parsed from the [text] with the hyphens in
+  /// pre-releases being parsed as a separate parts, if any.
+  ///
+  /// Example: `0.1.0-alpha.13-5-qwe` -> `preRelease: ['alpha', 13, 5, 'qwe']`.
+  ///
+  /// This is required due to [PubspecBuilder] using `git describe --tags`,
+  /// which returns hyphens instead of dots, and replacing that behavior seems a
+  /// bigger evil than this.
+  static Version parse(String text) {
+    final Version parsed = Version.parse(text);
+
+    return Version(
+      parsed.major,
+      parsed.minor,
+      parsed.patch,
+      pre: parsed.preRelease.isEmpty
+          ? null
+          : parsed.preRelease
+              .map((e) => e is String ? e.replaceAll('-', '.') : e)
+              .join('.'),
+      build: parsed.build.isEmpty ? null : parsed.build.join('.'),
+    );
+  }
```

Comment:
@tyranron, regarding the problem we discussed: `0.1.0-alpha.1-51-qweqweq` is HIGHER than `0.1.0-alpha.2`, despite our expectations, causing the update to be ignored. The same goes for any `0.1.0-alpha.X`, when compared to any `0.1.0-alpha.X-Y`.


### How so?

Imagine current version is `0.1.0-alpha.1-51-qweqweq`. And a new one is released: `0.1.0-alpha.2`: 
- `0.1.0-alpha.1-51-qweqweq` breaks down into `alpha` and `1-51-qweqweq`.
- `0.1.0-alpha.2` breaks down into `alpha` and `2`.

The first one happens due to [Semantic Version stating that pre-release identifiers must be divided by dots](https://semver.org/#spec-item-9):
> ... a series of dot separated identifiers

so `pub_semver` package recognizes the whole `1-51-qweqweq` as a single identifier. This causes to compare non-number with a number, and [Semantic Versioning states](https://semver.org/spec/v2.0.0.html#spec-item-11):
> 4. Numeric identifiers always have lower precedence than non-numeric identifiers.

So it results in `2` being LOWER than `1-51-qweqweq`.


### How to fix?

1) Replace hyphens with dots in every `$(git describe --tags)` - seems clumsy.
2) Use the parser I commented: it replaces the hyphens with dots in pre-releases. This will be used only for comparing the two versions in `UpgradeWorker` and nowhere else.